* Bio::Gadget

*Bio::Gadget* - Gadgets for bioinformatics

* Installation

: gem install bio-gadget

* Usage

To check all commands and the usages in this package

: bio-gadget help

Currently available commands are

- dedup :: Deduplicate fastq (via STDIN)
- demlt :: Demultiplex fastq by barcodes
- fqxz :: automatic (re)compression of *.fq(.gz|.bz2) files
- qvstat :: Statistics of quality values in *.qual file
- rgt2mtx :: Convert cuffdiff read group tracking file into tab-separated matrix
- wig5p :: Convert bam-format alignments into wig-format table
- wigchr :: Extract wiggle track on specified chromosome

** Memo                                                            :noexport:
- gtfensembl :: Create gtf of ENSEMBL Genes (a.k.a. ensGene.gtf) from XX files of UCSC annotation database; .txt.gz of 
- gtfucsc :: Create gtf of UCSC Genes from four (knownIsoforms.txt.gz, kgXref.txt.gz, kgTxInfo.txt.gz and knownGene.txt.gz) files of UCSC annotation database. The output gtf contains gene_id, transcript_id and gene_name attributes, so it convenients to use as a gene/transcript annotation file for tophat/cufflinks etc.

* Contributing

1. Fork it
2. Create your feature branch (`git checkout -b my-new-feature`)
3. (optional) Install gems (`bundle`)
4. Add your feature
5. Commit your changes (`git commit -am 'Added some feature'`)
6. Push to the branch (`git push origin my-new-feature`)
7. Create new Pull Request
